components/shared/embedder/lib.rs

```diff
@@ -585,3 +589,57 @@ pub enum LoadStatus {
     /// See <https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState>
     Complete,
 }
+
+/// <https://notifications.spec.whatwg.org/#notifications>
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Notification {
+    /// <https://notifications.spec.whatwg.org/#concept-title>
+    pub title: String,
+    /// <https://notifications.spec.whatwg.org/#body>
+    pub body: String,
+    /// <https://notifications.spec.whatwg.org/#tag>
+    pub tag: String,
+    /// <https://notifications.spec.whatwg.org/#concept-direction>
+    pub dir: NotificationDirection,
+    /// <https://notifications.spec.whatwg.org/#concept-language>
+    pub lang: String,
+    /// <https://notifications.spec.whatwg.org/#require-interaction-preference-flag>
+    pub require_interaction: bool,
+    /// <https://notifications.spec.whatwg.org/#silent-preference-flag>
+    pub silent: Option<bool>,
+    /// <https://notifications.spec.whatwg.org/#icon-url>
+    pub icon_url: Option<ServoUrl>,
+    /// <https://notifications.spec.whatwg.org/#icon-resource>
+    pub icon_resource: Option<Arc<Image>>,
+    /// <https://notifications.spec.whatwg.org/#badge-url>
+    pub badge_url: Option<ServoUrl>,
+    /// <https://notifications.spec.whatwg.org/#badge-resource>
+    pub badge_resource: Option<Arc<Image>>,
+    /// <https://notifications.spec.whatwg.org/#image-url>
+    pub image_url: Option<ServoUrl>,
+    /// <https://notifications.spec.whatwg.org/#image-resource>
+    pub image_resource: Option<Arc<Image>>,
+    /// <https://notifications.spec.whatwg.org/#actions>
```

Comment:
We shouldn't be sending raw images over IPC instead these should use IPCSharedMemory. If possible these should be the shared memory bits that are stored in the ImageCache.

Reply:
Thank you! This is my first time knowing that `IPCSharedMemory` exists.

Those `Arc<Image>` were cloned from an `ImageCache` response or an `ImageCache` cache result.
I found the [bytes of the Image](https://doc.servo.org/pixels/struct.Image.html#structfield.bytes) are `IpcSharedMemory`. If the `IpcSharedMemory` it used is what we expected, does it make sense to send an `Arc<Image>` or is it better to clone the shared memory bits to somewhere else?

Reply:
Ah, sorry, I didn't realize that! In that case I think it's fine to send the `Image`s over as-is. Instead of sending all of the bytes of the image, only the shared memory handle is sent.